internal/app/machined/pkg/controllers/runtime/events_sink.go

```diff
@@ -0,0 +1,190 @@
+	"github.com/rs/xid"
+	"github.com/talos-systems/go-procfs/procfs"
+	"github.com/talos-systems/siderolink/api/events"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/talos-systems/talos/internal/app/machined/pkg/runtime"
+	"github.com/talos-systems/talos/pkg/machinery/constants"
+	"github.com/talos-systems/talos/pkg/machinery/resources/network"
+)
+
+// EventsSinkController watches events and forwards them to the events sink server
+// if it's configured.
+type EventsSinkController struct {
+	V1Alpha1Events runtime.Watcher
+	Cmdline        *procfs.Cmdline
+	Drainer        *runtime.Drainer
+
+	drainSub *runtime.DrainSubscription
+	drain    bool
+	backlog  atomic.Int32
+	eventID  xid.ID
+}
+
+// Name implements controller.Controller interface.
+func (ctrl *EventsSinkController) Name() string {
+	return "v1alpha1.EventsSinkController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *EventsSinkController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: network.NamespaceName,
+			Type:      network.StatusType,
+			ID:        pointer.ToString(network.StatusID),
+			Kind:      controller.InputWeak,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *EventsSinkController) Outputs() []controller.Output {
+	return nil
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo,cyclop
+func (ctrl *EventsSinkController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) (err error) {
+	if ctrl.Cmdline == nil || ctrl.Cmdline.Get(constants.KernelParamEventsSink).First() == nil {
+		return nil
+	}
+
+	if ctrl.drainSub == nil {
+		ctrl.backlog.Store(-1)
+		ctrl.drainSub = ctrl.Drainer.Subscribe()
+	}
+
+	defer func() {
+		if ctrl.backlog.Load() == 0 {
+			ctrl.drainSub.Cancel()
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+		}
+
+		var netStatus resource.Resource
+
+		netStatus, err = r.Get(ctx, resource.NewMetadata(network.NamespaceName, network.StatusType, network.StatusID, resource.VersionUndefined))
+		if err != nil {
+			if state.IsNotFoundError(err) {
+				// no network state yet
+				continue
+			}
+
+			return fmt.Errorf("error reading network status: %w", err)
+		}
+
+		if !netStatus.(*network.Status).TypedSpec().AddressReady {
+			// wait for address
+			continue
+		}
+
+		break
+	}
+
+	errCh := make(chan error)
+
+	sink := ctrl.Cmdline.Get(constants.KernelParamEventsSink).First()
+
+	conn, err := grpc.DialContext(ctx, *sink, grpc.WithInsecure())
+	if err != nil {
+		return err
+	}
+
+	client := events.NewEventSinkServiceClient(conn)
+
+	opts := []runtime.WatchOptionFunc{}
+	if ctrl.eventID.IsNil() {
+		opts = append(opts, runtime.WithTailEvents(-1))
+	} else {
+		opts = append(opts, runtime.WithTailID(ctrl.eventID))
+	}
+
+	if err = ctrl.V1Alpha1Events.Watch(func(eventCh <-chan runtime.EventInfo) {
```

Comment:
nit: the func body could be extracted into a separate function for readability

Reply:
not sure about that... I don't really want to pass through all local variables into that function.
But can declare it as a func in the function body.

Reply:
I don't think there are _that_ many parameters? some of the are on the `ctrl` anyways, you probably need to pass just the client?